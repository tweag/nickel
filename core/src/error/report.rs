//! Error diagnostics reporting and serialization.
use super::*;

/// Serializable wrapper type to export diagnostics with a top-level attribute.
#[derive(serde::Serialize)]
pub struct DiagnosticsWrapper {
    pub diagnostics: Vec<Diagnostic<FileId>>,
}

impl From<Vec<Diagnostic<FileId>>> for DiagnosticsWrapper {
    fn from(diagnostics: Vec<Diagnostic<FileId>>) -> Self {
        Self { diagnostics }
    }
}

/// Available export formats for error diagnostics.
#[derive(Copy, Clone, Eq, PartialEq, Debug, Default, clap::ValueEnum)]
pub enum ErrorFormat {
    #[default]
    Text,
    Json,
    Yaml,
    Toml,
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub struct ColorOpt(pub(crate) clap::ColorChoice);

impl ColorOpt {
    fn for_terminal(self, is_terminal: bool) -> ColorChoice {
        match self.0 {
            clap::ColorChoice::Auto => {
                if is_terminal {
                    ColorChoice::Auto
                } else {
                    ColorChoice::Never
                }
            }
            clap::ColorChoice::Always => ColorChoice::Always,
            clap::ColorChoice::Never => ColorChoice::Never,
        }
    }
}

impl From<clap::ColorChoice> for ColorOpt {
    fn from(color_choice: clap::ColorChoice) -> Self {
        Self(color_choice)
    }
}

impl Default for ColorOpt {
    fn default() -> Self {
        Self(clap::ColorChoice::Auto)
    }
}

/// Pretty-print an error on stderr.
///
/// # Arguments
///
/// - `cache` is the file cache used during the evaluation, which is required by the reporting
/// infrastructure to point at specific locations and print snippets when needed.
pub fn report<E: IntoDiagnostics<FileId>>(
    cache: &mut Cache,
    error: E,
    format: ErrorFormat,
    color_opt: ColorOpt,
) {
    use std::io::{stderr, IsTerminal};

    let stdlib_ids = cache.get_all_stdlib_modules_file_id();
    report_with(
        &mut StandardStream::stderr(color_opt.for_terminal(stderr().is_terminal())).lock(),
        cache.files_mut(),
        stdlib_ids.as_ref(),
        error,
        format,
    )
}

/// Report an error on `stderr`, provided a file database and a list of stdlib file ids.
pub fn report_with<E: IntoDiagnostics<FileId>>(
    writer: &mut dyn WriteColor,
    files: &mut Files<String>,
    stdlib_ids: Option<&Vec<FileId>>,
    error: E,
    format: ErrorFormat,
) {
    let config = codespan_reporting::term::Config::default();
    let diagnostics = error.into_diagnostics(files, stdlib_ids);
    let stderr = std::io::stderr();

    let result = match format {
        ErrorFormat::Text => diagnostics.iter().try_for_each(|d| {
            codespan_reporting::term::emit(writer, &config, files, d).map_err(|err| err.to_string())
        }),
        ErrorFormat::Json => serde_json::to_writer(stderr, &DiagnosticsWrapper::from(diagnostics))
            .map(|_| eprintln!())
            .map_err(|err| err.to_string()),
        ErrorFormat::Yaml => serde_yaml::to_writer(stderr, &DiagnosticsWrapper::from(diagnostics))
            .map_err(|err| err.to_string()),
        ErrorFormat::Toml => toml::to_string(&DiagnosticsWrapper::from(diagnostics))
            .map(|repr| eprint!("{}", repr))
            .map_err(|err| err.to_string()),
    };

    match result {
        Ok(()) => (),
        Err(err) => panic!("error::report_with(): could not print an error on stderr: {err}"),
    };
}
