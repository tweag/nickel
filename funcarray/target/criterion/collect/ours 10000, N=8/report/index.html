<!DOCTYPE html>
<html>

<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <title>collect/ours 10000, N=8 - Criterion.rs</title>
    <style type="text/css">
        body {
            font: 14px Helvetica Neue;
            text-rendering: optimizelegibility;
        }

        .body {
            width: 960px;
            margin: auto;
        }

        th {
            font-weight: 200
        }

        th,
        td {
            padding-right: 3px;
            padding-bottom: 3px;
        }

        a:link {
            color: #1F78B4;
            text-decoration: none;
        }

        th.ci-bound {
            opacity: 0.6
        }

        td.ci-bound {
            opacity: 0.5
        }

        .stats {
            width: 80%;
            margin: auto;
            display: flex;
        }

        .additional_stats {
            flex: 0 0 60%
        }

        .additional_plots {
            flex: 1
        }

        h2 {
            font-size: 36px;
            font-weight: 300;
        }

        h3 {
            font-size: 24px;
            font-weight: 300;
        }

        #footer {
            height: 40px;
            background: #888;
            color: white;
            font-size: larger;
            font-weight: 300;
        }

        #footer a {
            color: white;
            text-decoration: underline;
        }

        #footer p {
            text-align: center
        }
    </style>
</head>

<body>
    <div class="body">
        <h2>collect/ours 10000, N=8</h2>
        <div class="absolute">
            <section class="plots">
                <table width="100%">
                    <tbody>
                        <tr>
                            <td>
                                <a href="pdf.svg">
                                    <img src="pdf_small.svg" alt="PDF of Slope" width="450" height="300" />
                                </a>
                            </td>
                            <td>
                                <a href="regression.svg">
                                    <img src="regression_small.svg" alt="Regression" width="450" height="300" />
                                </a>
                            </td>
                        </tr>
                    </tbody>
                </table>
            </section>
            <section class="stats">
                <div class="additional_stats">
                    <h4>Additional Statistics:</h4>
                    <table>
                        <thead>
                            <tr>
                                <th></th>
                                <th title="0.95 confidence level" class="ci-bound">Lower bound</th>
                                <th>Estimate</th>
                                <th title="0.95 confidence level" class="ci-bound">Upper bound</th>
                            </tr>
                        </thead>
                        <tbody>
                            <tr>
                                <td>Slope</td>
                                <td class="ci-bound">36.945 µs</td>
                                <td>37.154 µs</td>
                                <td class="ci-bound">37.388 µs</td>
                            </tr>
                            <tr>
                                <td>R&#xb2;</td>
                                <td class="ci-bound">0.9258223</td>
                                <td>0.9297691</td>
                                <td class="ci-bound">0.9248012</td>
                            </tr>
                            <tr>
                                <td>Mean</td>
                                <td class="ci-bound">36.859 µs</td>
                                <td>37.268 µs</td>
                                <td class="ci-bound">37.949 µs</td>
                            </tr>
                            <tr>
                                <td title="Standard Deviation">Std. Dev.</td>
                                <td class="ci-bound">523.26 ns</td>
                                <td>3.0170 µs</td>
                                <td class="ci-bound">5.0599 µs</td>
                            </tr>
                            <tr>
                                <td>Median</td>
                                <td class="ci-bound">36.656 µs</td>
                                <td>36.754 µs</td>
                                <td class="ci-bound">36.839 µs</td>
                            </tr>
                            <tr>
                                <td title="Median Absolute Deviation">MAD</td>
                                <td class="ci-bound">257.53 ns</td>
                                <td>351.29 ns</td>
                                <td class="ci-bound">450.97 ns</td>
                            </tr>
                        </tbody>
                    </table>
                </div>
                <div class="additional_plots">
                    <h4>Additional Plots:</h4>
                    <ul>
                        
                        <li>
                            <a href="typical.svg">Typical</a>
                        </li>
                        <li>
                            <a href="mean.svg">Mean</a>
                        </li>
                        <li>
                            <a href="SD.svg">Std. Dev.</a>
                        </li>
                        <li>
                            <a href="median.svg">Median</a>
                        </li>
                        <li>
                            <a href="MAD.svg">MAD</a>
                        </li>
                        <li>
                            <a href="slope.svg">Slope</a>
                        </li>
                    </ul>
                </div>
            </section>
            <section class="explanation">
                <h4>Understanding this report:</h4>
                <p>The plot on the left displays the average time per iteration for this benchmark. The shaded region
                    shows the estimated probability of an iteration taking a certain amount of time, while the line
                    shows the mean. Click on the plot for a larger view showing the outliers.</p>
                <p>The plot on the right shows the linear regression calculated from the measurements. Each point
                    represents a sample, though here it shows the total time for the sample rather than time per
                    iteration. The line is the line of best fit for these measurements.</p>
                <p>See <a href="https://bheisler.github.io/criterion.rs/book/user_guide/command_line_output.html#additional-statistics">the
                        documentation</a> for more details on the additional statistics.</p>
            </section>
        </div>
    </div>
    <div id="footer">
        <p>This report was generated by
            <a href="https://github.com/bheisler/criterion.rs">Criterion.rs</a>, a statistics-driven benchmarking
            library in Rust.</p>
    </div>
</body>

</html>