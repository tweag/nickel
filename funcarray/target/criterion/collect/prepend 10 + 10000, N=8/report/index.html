<!DOCTYPE html>
<html>

<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <title>collect/prepend 10 + 10000, N=8 - Criterion.rs</title>
    <style type="text/css">
        body {
            font: 14px Helvetica Neue;
            text-rendering: optimizelegibility;
        }

        .body {
            width: 960px;
            margin: auto;
        }

        th {
            font-weight: 200
        }

        th,
        td {
            padding-right: 3px;
            padding-bottom: 3px;
        }

        a:link {
            color: #1F78B4;
            text-decoration: none;
        }

        th.ci-bound {
            opacity: 0.6
        }

        td.ci-bound {
            opacity: 0.5
        }

        .stats {
            width: 80%;
            margin: auto;
            display: flex;
        }

        .additional_stats {
            flex: 0 0 60%
        }

        .additional_plots {
            flex: 1
        }

        h2 {
            font-size: 36px;
            font-weight: 300;
        }

        h3 {
            font-size: 24px;
            font-weight: 300;
        }

        #footer {
            height: 40px;
            background: #888;
            color: white;
            font-size: larger;
            font-weight: 300;
        }

        #footer a {
            color: white;
            text-decoration: underline;
        }

        #footer p {
            text-align: center
        }
    </style>
</head>

<body>
    <div class="body">
        <h2>collect/prepend 10 + 10000, N=8</h2>
        <div class="absolute">
            <section class="plots">
                <table width="100%">
                    <tbody>
                        <tr>
                            <td>
                                <a href="pdf.svg">
                                    <img src="pdf_small.svg" alt="PDF of Slope" width="450" height="300" />
                                </a>
                            </td>
                            <td>
                                <a href="regression.svg">
                                    <img src="regression_small.svg" alt="Regression" width="450" height="300" />
                                </a>
                            </td>
                        </tr>
                    </tbody>
                </table>
            </section>
            <section class="stats">
                <div class="additional_stats">
                    <h4>Additional Statistics:</h4>
                    <table>
                        <thead>
                            <tr>
                                <th></th>
                                <th title="0.95 confidence level" class="ci-bound">Lower bound</th>
                                <th>Estimate</th>
                                <th title="0.95 confidence level" class="ci-bound">Upper bound</th>
                            </tr>
                        </thead>
                        <tbody>
                            <tr>
                                <td>Slope</td>
                                <td class="ci-bound">224.06 ns</td>
                                <td>227.61 ns</td>
                                <td class="ci-bound">230.87 ns</td>
                            </tr>
                            <tr>
                                <td>R&#xb2;</td>
                                <td class="ci-bound">0.7139200</td>
                                <td>0.7250745</td>
                                <td class="ci-bound">0.7157160</td>
                            </tr>
                            <tr>
                                <td>Mean</td>
                                <td class="ci-bound">203.73 ns</td>
                                <td>209.17 ns</td>
                                <td class="ci-bound">214.65 ns</td>
                            </tr>
                            <tr>
                                <td title="Standard Deviation">Std. Dev.</td>
                                <td class="ci-bound">24.213 ns</td>
                                <td>28.089 ns</td>
                                <td class="ci-bound">31.885 ns</td>
                            </tr>
                            <tr>
                                <td>Median</td>
                                <td class="ci-bound">209.15 ns</td>
                                <td>215.59 ns</td>
                                <td class="ci-bound">220.81 ns</td>
                            </tr>
                            <tr>
                                <td title="Median Absolute Deviation">MAD</td>
                                <td class="ci-bound">17.471 ns</td>
                                <td>23.866 ns</td>
                                <td class="ci-bound">31.463 ns</td>
                            </tr>
                        </tbody>
                    </table>
                </div>
                <div class="additional_plots">
                    <h4>Additional Plots:</h4>
                    <ul>
                        
                        <li>
                            <a href="typical.svg">Typical</a>
                        </li>
                        <li>
                            <a href="mean.svg">Mean</a>
                        </li>
                        <li>
                            <a href="SD.svg">Std. Dev.</a>
                        </li>
                        <li>
                            <a href="median.svg">Median</a>
                        </li>
                        <li>
                            <a href="MAD.svg">MAD</a>
                        </li>
                        <li>
                            <a href="slope.svg">Slope</a>
                        </li>
                    </ul>
                </div>
            </section>
            <section class="explanation">
                <h4>Understanding this report:</h4>
                <p>The plot on the left displays the average time per iteration for this benchmark. The shaded region
                    shows the estimated probability of an iteration taking a certain amount of time, while the line
                    shows the mean. Click on the plot for a larger view showing the outliers.</p>
                <p>The plot on the right shows the linear regression calculated from the measurements. Each point
                    represents a sample, though here it shows the total time for the sample rather than time per
                    iteration. The line is the line of best fit for these measurements.</p>
                <p>See <a href="https://bheisler.github.io/criterion.rs/book/user_guide/command_line_output.html#additional-statistics">the
                        documentation</a> for more details on the additional statistics.</p>
            </section>
        </div>
        <section class="plots">
            <h3>Change Since Previous Benchmark</h3>
            <div class="relative">
                <table width="100%">
                    <tbody>
                        <tr>
                            <td>
                                <a href="both/pdf.svg">
                                    <img src="relative_pdf_small.svg" alt="PDF Comparison" width="450"
                                        height="300" />
                                </a>
                            </td>
                            <td>
                                <a href="both/regression.svg">
                                    <img src="relative_regression_small.svg" alt="Regression Comparison" width="450"
                                        height="300" />
                                </a>
                            </td>
                        </tr>
                    </tbody>
                </table>
            </div>
        </section>
        <section class="stats">
            <div class="additional_stats">
                <h4>Additional Statistics:</h4>
                <table>
                    <thead>
                        <tr>
                            <th></th>
                            <th title="0.95 confidence level" class="ci-bound">Lower bound</th>
                            <th>Estimate</th>
                            <th title="0.95 confidence level" class="ci-bound">Upper bound</th>
                            <th></th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                            <td>Change in time</td>
                            <td class="ci-bound">+5.3473%</td>
                            <td>+9.4480%</td>
                            <td class="ci-bound">+13.503%</td>
                            <td>(p = 0.00 &lt;
                                0.05)</td>
                        </tr>
                    </tbody>
                </table>
                Performance has regressed.
            </div>
            <div class="additional_plots">
                <h4>Additional Plots:</h4>
                <ul>
                    
                    <li>
                        <a href="change/mean.svg">Change in mean</a>
                    </li>
                    <li>
                        <a href="change/median.svg">Change in median</a>
                    </li>
                    <li>
                        <a href="change/t-test.svg">T-Test</a>
                    </li>
                </ul>
            </div>
        </section>
        <section class="explanation">
            <h4>Understanding this report:</h4>
            <p>The plot on the left shows the probability of the function taking a certain amount of time. The red
                curve represents the saved measurements from the last time this benchmark was run, while the blue curve
                shows the measurements from this run. The lines represent the mean time per iteration. Click on the
                plot for a larger view.</p>
            <p>The plot on the right shows the two regressions. Again, the red line represents the previous measurement
                while the blue line shows the current measurement.</p>
            <p>See <a href="https://bheisler.github.io/criterion.rs/book/user_guide/command_line_output.html#change">the
                    documentation</a> for more details on the additional statistics.</p>
        </section>
    </div>
    <div id="footer">
        <p>This report was generated by
            <a href="https://github.com/bheisler/criterion.rs">Criterion.rs</a>, a statistics-driven benchmarking
            library in Rust.</p>
    </div>
</body>

</html>