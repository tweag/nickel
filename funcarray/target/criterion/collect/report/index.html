<!DOCTYPE html>
<html>

<head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8">
    <title>collect Summary - Criterion.rs</title>
    <style type="text/css">
        body {
            font: 14px Helvetica Neue;
            text-rendering: optimizelegibility;
        }

        .body {
            width: 960px;
            margin: auto;
        }

        a:link {
            color: #1F78B4;
            text-decoration: none;
        }

        h2 {
            font-size: 36px;
            font-weight: 300;
        }

        h3 {
            font-size: 24px;
            font-weight: 300;
        }

        #footer {
            height: 40px;
            background: #888;
            color: white;
            font-size: larger;
            font-weight: 300;
        }

        #footer a {
            color: white;
            text-decoration: underline;
        }

        #footer p {
            text-align: center
        }
    </style>
</head>

<body>
    <div class="body">
        <h2>collect</h2>
        <h3>Violin Plot</h3>
        <a href="violin.svg">
            <img src="violin.svg" alt="Violin Plot" />
        </a>
        <p>This chart shows the relationship between function/parameter and iteration time. The thickness of the shaded
            region indicates the probability that a measurement of the given function/parameter would take a particular
            length of time.</p>
        <section class="plots">
            <a href="../../collect/slice 100 out of 10k, N=64/report/index.html">
                <h4>collect/slice 100 out of 10k, N=64</h4>
            </a>
            <table width="100%">
                <tbody>
                    <tr>
                        <td>
                            <a href="../../collect/slice 100 out of 10k, N=64/report/pdf.svg">
                                <img src="../../collect/slice 100 out of 10k, N=64/report/pdf_small.svg" alt="PDF of Slope" width="450"
                                    height="300" />
                            </a>
                        </td>
                        <td>
                            <a href="../../collect/slice 100 out of 10k, N=64/report/regression.svg">
                                <img src="../../collect/slice 100 out of 10k, N=64/report/regression_small.svg" alt="Regression" width="450"
                                    height="300" />
                            </a>
                        </td>
                    </tr>
                </tbody>
            </table>
        </section>
        <section class="plots">
            <a href="../../collect/slice 100 out of 10k, N=8/report/index.html">
                <h4>collect/slice 100 out of 10k, N=8</h4>
            </a>
            <table width="100%">
                <tbody>
                    <tr>
                        <td>
                            <a href="../../collect/slice 100 out of 10k, N=8/report/pdf.svg">
                                <img src="../../collect/slice 100 out of 10k, N=8/report/pdf_small.svg" alt="PDF of Slope" width="450"
                                    height="300" />
                            </a>
                        </td>
                        <td>
                            <a href="../../collect/slice 100 out of 10k, N=8/report/regression.svg">
                                <img src="../../collect/slice 100 out of 10k, N=8/report/regression_small.svg" alt="Regression" width="450"
                                    height="300" />
                            </a>
                        </td>
                    </tr>
                </tbody>
            </table>
        </section>
        <section class="plots">
            <a href="../../collect/slice 100 out of 1k, N=64/report/index.html">
                <h4>collect/slice 100 out of 1k, N=64</h4>
            </a>
            <table width="100%">
                <tbody>
                    <tr>
                        <td>
                            <a href="../../collect/slice 100 out of 1k, N=64/report/pdf.svg">
                                <img src="../../collect/slice 100 out of 1k, N=64/report/pdf_small.svg" alt="PDF of Slope" width="450"
                                    height="300" />
                            </a>
                        </td>
                        <td>
                            <a href="../../collect/slice 100 out of 1k, N=64/report/regression.svg">
                                <img src="../../collect/slice 100 out of 1k, N=64/report/regression_small.svg" alt="Regression" width="450"
                                    height="300" />
                            </a>
                        </td>
                    </tr>
                </tbody>
            </table>
        </section>
        <section class="plots">
            <a href="../../collect/slice 100 out of 1k, N=8/report/index.html">
                <h4>collect/slice 100 out of 1k, N=8</h4>
            </a>
            <table width="100%">
                <tbody>
                    <tr>
                        <td>
                            <a href="../../collect/slice 100 out of 1k, N=8/report/pdf.svg">
                                <img src="../../collect/slice 100 out of 1k, N=8/report/pdf_small.svg" alt="PDF of Slope" width="450"
                                    height="300" />
                            </a>
                        </td>
                        <td>
                            <a href="../../collect/slice 100 out of 1k, N=8/report/regression.svg">
                                <img src="../../collect/slice 100 out of 1k, N=8/report/regression_small.svg" alt="Regression" width="450"
                                    height="300" />
                            </a>
                        </td>
                    </tr>
                </tbody>
            </table>
        </section>
    </div>
    <div id="footer">
        <p>This report was generated by
            <a href="https://github.com/bheisler/criterion.rs">Criterion.rs</a>, a statistics-driven benchmarking
            library in Rust.</p>
    </div>
</body>

</html>