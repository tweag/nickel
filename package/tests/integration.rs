use libtest_mimic::{Arguments, Trial};
use std::{
    path::Path,
    process::{Command, ExitCode},
    sync::Arc,
};
use util::publish_package;

use nickel_lang_core::error::report::report_as_str;
use nickel_lang_package::{
    config::Config, index::PackageIndex, lock::LockFile, manifest::MANIFEST_NAME, resolve,
    snapshot::Snapshot, ManifestFile,
};
use nickel_lang_utils::project_root::project_root;
use tempfile::TempDir;

mod util;

pub fn main() -> ExitCode {
    let args = Arguments::from_args();
    let root = project_root();
    let root_str = project_root().into_os_string().into_string().unwrap();

    let manifest_glob = glob::glob(&format!(
        "{root_str}/package/tests/integration/inputs/path/**/Nickel-pkg.ncl"
    ))
    .unwrap();

    let fixture = Arc::new(Fixture::default());

    let tests: Vec<_> = manifest_glob
        .map(|p| {
            let path = p.unwrap();
            let name = path.strip_prefix(&root).unwrap().to_owned();
            let fixture = fixture.clone();
            Trial::test(name.display().to_string(), move || {
                generate_lock_file(&name, &fixture.config);
                Ok(())
            })
        })
        .collect();

    libtest_mimic::run(&args, tests).exit_code()
}

macro_rules! assert_lock_snapshot_filtered {
    { $name:expr, $snapshot:expr } => {
        insta::with_settings!({filters => vec![
            // Lock files contain git ids, and I haven't figured out if it's possible to
            // get them consistent between runs (e.g., they include hashes of timestamps).
            // So we just filter them out of the comparison.
            (r#""id": "[a-z0-9]+""#, r#""id": <GENERATED>"#)
        ]},
        {
            insta::assert_snapshot!($name, $snapshot);
        })
    }
}

macro_rules! assert_snapshot_filtered {
    { $name:expr, $snapshot:expr } => {
        insta::with_settings!({filters => vec![
            // Error messages contain paths, and windows displays them differently
            (r#"'Path [^ ]+"#, r#"'Path <filtered>"#)
        ]},
        {
            insta::assert_snapshot!($name, $snapshot);
        })
    }
}

// We'd like to test git dependencies, but it's considered bad form (and is annoying to manage)
// to nest the test git repos in our main repo. So what we do is just keep the contents of our
// test git repos in `package/tests/integration/inputs/git`. Then when we run our tests, we
// create temporary git repos for these contents, and use the source replacement mechanism
// to redirect to these temporary git repos.
struct Fixture {
    _tmp_dir: TempDir,
    _git_repos: TempDir,
    config: Config,
}

impl Default for Fixture {
    fn default() -> Self {
        let (tmp_dir, mut config) = util::test_config();
        let git_repos = TempDir::new().unwrap();

        set_up_git_repos(&mut config, &git_repos);
        set_up_test_index(&config, git_repos.path());

        Fixture {
            _tmp_dir: tmp_dir,
            _git_repos: git_repos,
            config,
        }
    }
}

// Creates git repos and populates them with files from our integration test suite.
//
// Modifies the provided config so that the git dependency
// `https://example.com/my-git-repo` will be redirected a git repo with the
// contents of `package/tests/integration/inputs/git/my-git-repo`
fn set_up_git_repos(config: &mut Config, git_dir: &TempDir) {
    let git_inputs =
        std::fs::read_dir(project_root().join("package/tests/integration/inputs/git")).unwrap();

    for input in git_inputs {
        let input = input.unwrap();
        let input_path = input.path();
        let file_name = input_path.file_name().unwrap();

        let dir_path = git_dir.path().join(file_name);

        set_up_git_repo(&input_path, &dir_path);
        let orig_url = gix::Url::try_from(format!(
            "https://example.com/{}",
            Path::new(file_name).display()
        ))
        .unwrap();
        let new_url = gix::Url::try_from(dir_path.display().to_string()).unwrap();
        config.git_replacements.insert(orig_url, new_url);
    }
}

// Copies the directory `contents` to `to`, and initializes a git repo in the
// new location.
fn set_up_git_repo(contents: &Path, to: &Path) {
    let run = |cmd: &mut Command| {
        let output = cmd.output().unwrap();
        assert!(
            output.status.success(),
            "command {cmd:?} failed, stdout {}, stderr {}",
            String::from_utf8_lossy(&output.stdout),
            String::from_utf8_lossy(&output.stderr)
        );
    };

    let run_in_dir = |cmd: &mut Command| {
        run(cmd.current_dir(to));
    };

    // The rust stdlib doesn't have anything for recursively copying a directory. There are
    // some crates for that, but it's easier just to shell out.
    run(Command::new("cp").arg("-r").arg(contents).arg(to));

    // We have some hacky ways to test branch/tag fetching: if the input contains a tag.txt file,
    // make a git tag named with the contents of that file. If the input contains a branch.txt file,
    // make a git branch named with the contents of that file.
    let tag = std::fs::read_to_string(to.join("tag.txt")).ok();
    let branch = std::fs::read_to_string(to.join("branch.txt")).ok();

    run_in_dir(Command::new("git").arg("init"));
    run_in_dir(Command::new("git").args(["config", "user.email", "me@example.com"]));
    run_in_dir(Command::new("git").args(["config", "user.name", "me"]));

    if let Some(branch) = branch {
        run_in_dir(Command::new("git").args(["commit", "-m", "initial", "--allow-empty"]));
        run_in_dir(Command::new("git").args(["checkout", "-b", branch.trim()]));
    }

    run_in_dir(Command::new("git").args(["add", "--all"]));
    run_in_dir(Command::new("git").args(["commit", "-m", "initial"]));

    if let Some(tag) = tag {
        run_in_dir(Command::new("git").args(["tag", tag.trim()]));
    }
}

// Creates an index in the configured location (which must be a directory), and
// populates it with the packages found in `package/tests/integration/inputs/index`.
fn set_up_test_index(config: &Config, tmp_dir: &Path) {
    // The packages for populating our index stored in the form
    // package/tests/integration/inputs/index/github/<org>/<package>/<version>.
    let index_dir =
        std::fs::read_dir(project_root().join("package/tests/integration/inputs/index/github"))
            .unwrap();
    for org in index_dir {
        let org = org.unwrap();
        for pkg in std::fs::read_dir(org.path()).unwrap() {
            let pkg = pkg.unwrap();
            for version in std::fs::read_dir(pkg.path()).unwrap() {
                let version = version.unwrap();

                // Package publishing needs a git repo to get a hash from, so first
                // copy the contents into a git repo.
                let tmp_location = tmp_dir
                    .join("index-git-tmp")
                    .join(org.file_name())
                    .join(pkg.file_name())
                    .join(version.file_name());
                std::fs::create_dir_all(tmp_location.parent().unwrap()).unwrap();
                set_up_git_repo(&version.path(), &tmp_location);

                let manifest_path = tmp_location.join(MANIFEST_NAME);
                let manifest = ManifestFile::from_path(&manifest_path).unwrap();
                assert_eq!(
                    manifest.version,
                    version.file_name().into_string().unwrap().parse().unwrap()
                );
                publish_package(
                    config,
                    &manifest,
                    &format!(
                        "github:{}/{}",
                        org.file_name().into_string().unwrap(),
                        pkg.file_name().into_string().unwrap()
                    ),
                );
            }
        }
    }
}

fn generate_lock_file(path: &Path, config: &Config) {
    let full_path = project_root().join(path);
    let index_dir = TempDir::new().unwrap();

    let config = config.clone();

    // Make an empty git repo as the index.
    Command::new("git")
        .arg("init")
        .current_dir(index_dir.path())
        .output()
        .unwrap();
    Command::new("git")
        .args(["config", "user.email", "me@example.com"])
        .current_dir(index_dir.path())
        .output()
        .unwrap();
    Command::new("git")
        .args(["config", "user.name", "Me"])
        .current_dir(index_dir.path())
        .output()
        .unwrap();

    Command::new("git")
        .args(["commit", "--allow-empty", "-m", "initial"])
        .current_dir(index_dir.path())
        .output()
        .unwrap();

    let manifest = match ManifestFile::from_path(&full_path) {
        Ok(m) => m,
        Err(nickel_lang_package::error::Error::ManifestEval {
            package: _package,
            mut files,
            error,
        }) => {
            panic!("{}", report_as_str(&mut files, error, Default::default()));
        }
        Err(e) => panic!("{}", e),
    };
    let index = PackageIndex::shared_or_initialize(config.clone()).unwrap();

    let snap = Snapshot::new(&config, &manifest.parent_dir, &manifest).unwrap();
    match resolve::resolve(&manifest, snap, index, config) {
        Ok(resolution) => {
            let lock = LockFile::new(&manifest, &resolution).unwrap();
            let lock_contents = serde_json::to_string_pretty(&lock).unwrap();

            assert_lock_snapshot_filtered!(path.display().to_string(), lock_contents);
        }
        Err(e) => {
            assert_snapshot_filtered!(path.display().to_string(), e.to_string());
        }
    }
}
