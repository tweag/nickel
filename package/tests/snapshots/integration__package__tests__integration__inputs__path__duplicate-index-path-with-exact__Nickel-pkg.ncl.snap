---
source: package/tests/integration.rs
expression: lock_contents
---
{
  "dependencies": {
    "dep": {
      "name": "dep"
    },
    "leaf": {
      "name": "leaf"
    }
  },
  "packages": {
    "dep": {
      "precise": "Path",
      "dependencies": {
        "leaf": {
          "name": "leaf"
        }
      }
    },
    "leaf": {
      "precise": {
        "Index": {
          "id": {
            "type": "github",
            "org": "example",
            "name": "leaf"
          },
          "version": {
            "major": 0,
            "minor": 1,
            "patch": 1,
            "pre": ""
          }
        }
      },
      "dependencies": {}
    }
  }
}
